query-algebrizer/src/clauses/pattern.rs

```diff
@@ -263,15 +264,72 @@ impl ConjoiningClauses {
         }
     }
 
-    pub fn apply_pattern<'s, 'p>(&mut self, schema: &'s Schema, pattern: Pattern) {
+    // TODO: generalize.
+    fn attempt_cache_lookup(&mut self, known: Known, pattern: &Pattern) -> bool {
+        // Precondition: default source. If it's not default, don't call this.
+        assert!(pattern.has_default_source());
+
+        let schema = known.schema;
+
+        if pattern.tx != PatternNonValuePlace::Placeholder {
+            return false;
+        }
+
+        // See if we can use the cache.
+        match pattern.attribute {
+            PatternNonValuePlace::Entid(attr) => {
+                if !schema.is_attribute(attr) {
+                    // Furthermore, that entid must resolve to an attribute. If it doesn't, this
+                    // query is meaningless.
+                    self.mark_known_empty(EmptyBecause::InvalidAttributeEntid(attr));
+                    return true;
+                }
+
+                if known.is_attribute_cached(attr) &&
+                   pattern.tx == PatternNonValuePlace::Placeholder {
+                    if let PatternNonValuePlace::Entid(entity) = pattern.entity {
+                        match pattern.value {
+                            PatternValuePlace::Variable(ref var) => {
+                                match known.get_values_for_entid(attr, entity) {
+                                    // For thoroughness we'd like to check for len == 0, but
+                                    // we shouldn't need to do that.
+                                    None => {
+                                        self.mark_known_empty(EmptyBecause::CachedAttributeHasNoValues {
+                                            entity: entity,
+                                            attr: attr,
+                                        });
+                                        return true;
+                                    },
+                                    Some(items) if items.len() == 1 => {
+                                        println!("{} is known to be {:?}", var, items[0]);
+                                        self.bind_value(var, items[0].clone());
+                                        return true;
+                                    }
+                                    _ => {},    // Nothing we can do.
+                                }
+                            }
+                            _ => {},      // TODO: check constant values against cache.
+                        }
+                    }
+                }
+            },
+            _ => {},
```

Comment:
If I channel @ncalexan then we should be enumerating all of our options here, but probably not worth it.